controllers/storagecluster/odfinfoconfig.go

```diff
@@ -0,0 +1,248 @@
+	OdfDeploymentTypeExternal = "external"
+	OdfDeploymentTypeInternal = "internal"
+	RookCephMonSecretName     = "rook-ceph-mon"
+	FsidKey                   = "fsid"
+	OdfOperatorNamePrefix     = "odf-operator"
+	OdfInfoConfigMapName      = "odf-info"
+	OdfInfoMapKind            = "ConfigMap"
+)
+
+type odfInfoConfig struct{}
+
+var mutex sync.RWMutex
+
+// ensureCreated ensures that a ConfigMap resource exists with its Spec in
+// the desired state.
+func (obj *odfInfoConfig) ensureCreated(r *StorageClusterReconciler, sc *ocsv1.StorageCluster) (reconcile.Result, error) {
+	operatorNamespace, err := util.GetOperatorNamespace()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	odfInfoConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      OdfInfoConfigMapName,
+			Namespace: operatorNamespace,
+		},
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+	_, err = ctrl.CreateOrUpdate(r.ctx, r.Client, odfInfoConfigMap, func() error {
+		if err := ctrl.SetControllerReference(sc, odfInfoConfigMap, r.Scheme); err != nil {
+			return err
+		}
+		r.Log.Info("Creating or updating odf-info configmap", OdfInfoMapKind, klog.KRef(sc.Namespace, OdfInfoConfigMapName))
+		odfInfoData, configErr := getOdfInfoData(r, sc)
+		if configErr != nil {
+			return fmt.Errorf("failed to get ODF info config data: %w", configErr)
+		}
+		odfInfoKey := obj.getOdfInfoKeyName(sc)
+		// Creates or appends to the data map
+		odfInfoConfigMap.Data[odfInfoKey] = odfInfoData
+		return nil
+	})
+	if err != nil {
+		r.Log.Error(err, "failed to create or update odf-info config", OdfInfoMapKind, klog.KRef(sc.Namespace, OdfInfoConfigMapName))
+		return reconcile.Result{}, fmt.Errorf("failed to create or update odf-info config: %w", err)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (obj *odfInfoConfig) getOdfInfoKeyName(sc *ocsv1.StorageCluster) string {
+	return sc.Namespace + "/" + sc.Name + "." + OdfInfoKeyName
+}
+
+// ensureDeleted is dummy func for the odfInfoConfig
+func (obj *odfInfoConfig) ensureDeleted(r *StorageClusterReconciler, sc *ocsv1.StorageCluster) (reconcile.Result, error) {
+	operatorNamespace, err := util.GetOperatorNamespace()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	var odfInfoConfigMap corev1.ConfigMap
+	if err = r.Client.Get(r.ctx, types.NamespacedName{Namespace: operatorNamespace}, &odfInfoConfigMap); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+	if len(odfInfoConfigMap.Data) > 1 {
+		delete(odfInfoConfigMap.Data, obj.getOdfInfoKeyName(sc))
+		if err = r.Client.Update(r.ctx, &odfInfoConfigMap); err != nil {
+			r.Log.Error(err, "Failed to delete key in odf-info.", "ConfigMap", klog.KRef(odfInfoConfigMap.Namespace, odfInfoConfigMap.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to delete key %v in odf-info %v: %v", obj.getOdfInfoKeyName(sc), odfInfoConfigMap.Name, err)
+		}
+	} else {
+		if err = r.Client.Delete(r.ctx, &odfInfoConfigMap); err != nil {
```

Comment:
I think you might want to skip `NotFound` error.

Reply:
Don't we have `r.delete` that already does that?

Reply:
I'm catching it prior to this line, right at the beginning of the function, when I'm trying to retrieve the resource.

Reply:
Okay, so something like should do I think
```go
if err = r.Client.Delete(r.ctx, odfInfoConfigMap); err != nil && !errors.IsNotFound(err) {
    //something
}
```

Reply:
Resolving as per @leelavg's approval